src/main/java/com/cloudbees/jenkins/support/filter/FilteredOutputStream.java

```diff
@@ -0,0 +1,202 @@
+ * written is filtered.
+ *
+ * @see ContentFilter
+ * @see FilteredWriter
+ * @since TODO
+ */
+public class FilteredOutputStream extends FilterOutputStream {
+
+    static final Pattern EOL = Pattern.compile("\r?\n");
+    static final int DEFAULT_DECODER_CAPACITY = 1024;
+    private static final String UNKNOWN_INPUT = "\uFFFD";
+
+    @GuardedBy("this")
+    private final ByteBuffer in = ByteBuffer.allocate(256);
+    @GuardedBy("this")
+    private CharBuffer buf = CharBuffer.allocate(DEFAULT_DECODER_CAPACITY);
+    private final Charset charset;
+    @GuardedBy("this")
+    private final CharsetDecoder decoder;
+    private final ContentFilter contentFilter;
+
+    /**
+     * Constructs a filtered stream using the provided filter and assuming UTF-8.
+     *
+     * @param out           output stream to write filtered content to
+     * @param contentFilter content filter to apply to lines written through this stream
+     */
+    public FilteredOutputStream(@Nonnull OutputStream out, @Nonnull ContentFilter contentFilter) {
+        this(out, StandardCharsets.UTF_8, contentFilter);
+    }
+
+    /**
+     * Constructs a filtered stream using the provided filter and charset.
+     *
+     * @param out           output stream to write filtered content to
+     * @param charset       character set to use for decoding and encoding bytes written to this stream
+     * @param contentFilter content filter to apply to lines written through this stream
+     */
+    public FilteredOutputStream(@Nonnull OutputStream out, @Nonnull Charset charset, @Nonnull ContentFilter contentFilter) {
+        super(out);
+        this.charset = charset;
+        this.decoder = charset.newDecoder()
+                .onMalformedInput(CodingErrorAction.REPLACE)
+                .onUnmappableCharacter(CodingErrorAction.REPLACE)
+                .replaceWith(UNKNOWN_INPUT);
+        this.contentFilter = contentFilter;
+    }
+
+    @Override
+    public synchronized void write(int b) throws IOException {
+        write(new byte[]{(byte) b}, 0, 1);
+    }
+
+    @Override
+    public synchronized void write(@Nonnull byte[] b) throws IOException {
+        write(b, 0, b.length);
+    }
+
+    @Override
+    public synchronized void write(@Nonnull byte[] b, int off, int len) throws IOException {
+        while (len > 0) {
+            int toCopy = Math.min(in.remaining(), len);
+            if (toCopy == 0) throw new IllegalStateException();
+            in.put(b, off, toCopy);
+            decodeFilterFlushLines(false);
+            len -= toCopy;
+            off += toCopy;
+        }
+        filterFlushLines();
+    }
+
+    /**
+     * Flushes the current buffered contents and filters them as is.
+     */
+    @Override
+    public synchronized void flush() throws IOException {
+        if (buf.position() > 0) {
+            buf.flip();
+            String contents = buf.toString();
```

Comment:
Couldn't this be incorrect if a component calls flush when it has not written a full line? I would use `filterFlushLines` but leave any partially buffered lines buffered.

Reply:
The point here is to force a filtering when a flush is called. This could be in files that don't end a new line, for example.

Reply:
Yes, but the case where files don't end in a new line is always handled by `#close`, right?

Reply:
Files don't necessarily get closed while writing them to the zip file. That would need to be forced I suppose.

Reply:
Yes I was confused here thinking that we call `FilterOutputStream#close` after `content.writeTo(filteredOut)` in SupportPlugin, but that's not the case.